Modules/mathmodule.c

```diff
@@ -1114,15 +1114,17 @@ math_ceil(PyObject *module, PyObject *number)
     _Py_IDENTIFIER(__ceil__);
     PyObject *method, *result;
 
-    method = _PyObject_LookupSpecial(number, &PyId___ceil__);
-    if (method == NULL) {
+    if (!PyFloat_CheckExact(number)) {
+        method = _PyObject_LookupSpecial(number, &PyId___ceil__);
+        if (method != NULL) {
+            result = _PyObject_CallNoArg(method);
+            Py_DECREF(method);
+            return result;
+        }
         if (PyErr_Occurred())
             return NULL;
-        return math_1_to_int(number, ceil, 0);
     }
-    result = _PyObject_CallNoArg(method);
-    Py_DECREF(method);
-    return result;
+    return math_1_to_int(number, ceil, 0);
```

Comment:
I think we should be able to dispense with the complications in `math_1_to_int` and simply use `PyLong_FromDouble(ceil(number))` here (and similarly for `floor`, of course). `floor` and `ceil` should never be setting `errno` or raising any error conditions: there aren't any circumstances where they can produce a NaN from non-NaN input, or an non-finite result from finite input.

Reply:
And of course there are badly-behaved libm implementations out there, but `floor` and `ceil` are quite hard to get wrong even for those.

Reply:
> PyLong_FromDouble(ceil(number))

Gah. There should be a `PyFloat_AsDouble` in there somewhere, of course. But you know what I meant. :-)

Reply:
Good idea!